core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java

```diff
@@ -500,29 +501,53 @@ private void recoverClusterStatus(int newViewId, final List<Address> clusterMemb
       latch.await(getGlobalTimeout(), TimeUnit.MILLISECONDS);
    }
 
-   public void updateCacheMembers(List<Address> newClusterMembers) {
+   public void updateCacheMembers(int viewId) {
       try {
-         log.tracef("Updating cluster members for all the caches. New list is %s", newClusterMembers);
-         try {
-            // If we get a SuspectException here, it means we will have a new view soon and we can ignore this one.
-            confirmMembersAvailable();
-         } catch (SuspectException e) {
-            log.tracef("Node %s left while updating cache members", e.getSuspect());
-            return;
-         }
-
-         for (ClusterCacheStatus cacheStatus : cacheStatusMap.values()) {
-            cacheStatus.doHandleClusterView();
-         }
+         // Confirm that view's members are all available first, so in a network split scenario
+         // we can enter degraded mode without starting a rebalance first
+         // We don't really need to run on the view handling executor because ClusterCacheStatus
+         // has its own synchronization
+         confirmMembersAvailable().whenCompleteAsync((ignored, throwable) -> {
+            try {
+               doUpdateCacheMembers(viewId, throwable);
+            } catch (Throwable t) {
+               if (clusterManagerStatus.isRunning()) {
+                  log.errorUpdatingMembersList(viewId, t);
+               }
+            }
+         }, viewHandlingExecutor);
       } catch (Exception e) {
          if (clusterManagerStatus.isRunning()) {
-            log.errorUpdatingMembersList(e);
+            log.errorUpdatingMembersList(viewId, e);
          }
       }
    }
 
-   private void confirmMembersAvailable() throws Exception {
-      transport.invokeRemotely(null, HeartBeatCommand.INSTANCE, ResponseMode.SYNCHRONOUS, getGlobalTimeout(), null, DeliverOrder.NONE, false);
+   public void doUpdateCacheMembers(int viewId, Throwable throwable) throws Throwable {
```

Comment:
Method seems a bit superfluous - Do you think if you consolidated it would look a bit nicer? I personally am not a big fan of using exceptions for flow control when we will catch it in the same class and ignore it.

Reply:
I was trying to keep the code that ignores exceptions in the same method, but you're right, it's not very pretty.